### PR TITLE
Hardcoded Trace Assertions rather than arbitrary functions

### PR DESCRIPTION
**Pull request by shmish111**
This means we can use them in the http api

Once this PR has been merged I will add assertions as endpoints to the http api and fit it into the Trace api.

### CODE REVIEW

File: wallet-api/src/Wallet/Emulator/Types.hs [hunk 1]
```diff
@@ -164,13 +163,37 @@ instance WalletAPI EmulatedWalletApi where
...
-    if notElem txn (join $ emChain emState) then Just $ AssertionError $ "Txn not validated: " <> T.pack (show txn) else Nothing
+assert :: (MonadEmulator m) => Assertion -> m ()
+assert (IsValidated txn) = isValidated txn
+assert (OwnFundsEqual wallet value) = ownFundsEqual wallet value
+
+ownFundsEqual :: (MonadEmulator m) => Wallet -> Value -> m ()
+ownFundsEqual wallet value = do
+  es <- get
+  ws <- case Map.lookup wallet $ emWalletState es of
+        Nothing -> throwError . AssertionError $ "Wallet not found"
```
Comment (vmchale) [inline wallet-api/src/Wallet/Emulator/Types.hs]: I'd prefer `throwError $ AssertionError "Wallet not found"` or just `throwError (AssertionError $ "Wallet not found")`.

Comment (vmchale) [inline wallet-api/src/Wallet/Emulator/Types.hs]: I'd prefer `throwError $ AssertionError "Wallet not found"` or just `throwError (AssertionError $ "Wallet not found")`.

File: wallet-api/src/Wallet/Emulator/Types.hs [hunk 2]
```diff
@@ -164,13 +163,37 @@ instance WalletAPI EmulatedWalletApi where
...
+  | OwnFundsEqual Wallet Value
+
 newtype AssertionError = AssertionError T.Text
     deriving Show
 
-isValidated :: Tx -> Assertion
-isValidated txn emState =
-    if notElem txn (join $ emChain emState) then Just $ AssertionError $ "Txn not validated: " <> T.pack (show txn) else Nothing
+assert :: (MonadEmulator m) => Assertion -> m ()
+assert (IsValidated txn) = isValidated txn
```
Comment (vmchale) [inline wallet-api/src/Wallet/Emulator/Types.hs]: This seems to be throwing off the `stylish-haskell` CI.

File: wallet-api/src/Wallet/Emulator/Types.hs (hunk 1 above)
Comment (shmish111) [reply wallet-api/src/Wallet/Emulator/Types.hs]: fixed

File: wallet-api/src/Wallet/Emulator/Types.hs (hunk 2 above)
Comment (shmish111) [reply wallet-api/src/Wallet/Emulator/Types.hs]: seems to be a duplicate.

Comment (shmish111) [reply wallet-api/src/Wallet/Emulator/Types.hs]: sorry, forget that, github broken UI mixing things up, this should also be fixed now

File: wallet-api/src/Wallet/Emulator/Types.hs [hunk 3]
```diff
@@ -309,21 +328,11 @@ blockchainActions = Op.singleton BlockchainActions
...
-
--- | Assert that the funds owned by a wallet are equal to an amount
 assertOwnFundsEq :: Wallet -> Value -> Trace ()
-assertOwnFundsEq w target = assertOwnFunds w p where
-    tshow :: Show a => a -> T.Text
-    tshow = T.pack . show
-    p v
-        | v == target = Nothing
-        | otherwise   = Just $ T.unwords ["Funds in wallet", tshow w, "were", tshow v, ". Expected:", tshow target]
+assertOwnFundsEq wallet value = assertion $ OwnFundsEqual wallet value
```
Comment (michaelpj) [inline wallet-api/src/Wallet/Emulator/Types.hs]: You might as well eta-reduce this like you did for the other one.

Comment (shmish111) [reply wallet-api/src/Wallet/Emulator/Types.hs]: done

**Comment by michaelpj** [comment]
`hlint` doesn't like `undefined`. Why not use the HTTP status code like you did for the other one?

**Comment by michaelpj** [comment]
Ah you dropped it, never mind.

**Pull request merged**